### ISSUE TITLE
CI builds fail with "does master exist?"

### ISSUE THREAD

**Issue by nategreen** [atlassian/changesets#517]
## Affected Packages

`@changesets/cli`
`@changesets/git`

## Problem

It seems that `2.13.0` of `@changesets/cli` broke our CI builds. (I'm assuming that's because of the update to `@changesets/git`.) We now get this when running `changeset status` on PRs:

```bash
🦋  error Error: Failed to find where HEAD diverged from master. Does master exist?
🦋  error     at getDivergedCommit (/opt/hostedtoolcache/node/12.20.1/x64/lib/node_modules/@changesets/cli/node_modules/@changesets/git/dist/git.cjs.dev.js:58:11)
🦋  error     at async getChangedFilesSince (/opt/hostedtoolcache/node/12.20.1/x64/lib/node_modules/@changesets/cli/node_modules/@changesets/git/dist/git.cjs.dev.js:169:22)
🦋  error     at async Object.getChangedPackagesSinceRef (/opt/hostedtoolcache/node/12.20.1/x64/lib/node_modules/@changesets/cli/node_modules/@changesets/git/dist/git.cjs.dev.js:208:24)
🦋  error     at async getStatus (/opt/hostedtoolcache/node/12.20.1/x64/lib/node_modules/@changesets/cli/dist/cli.cjs.dev.js:953:27)
🦋  error     at async run$1 (/opt/hostedtoolcache/node/12.20.1/x64/lib/node_modules/@changesets/cli/dist/cli.cjs.dev.js:1197:11)
```

We're running this on Ubuntu machines ("latest" version) in Azure Pipelines. Couldn't reproduce the problem locally.

Confirmed that installing `@changesets/cli` version `2.12.0` fixed the problem. 

## Proposed solution

🤷‍♂️ No clue. Let me know if there is other info I can provide to help debug. Thanks!

**Comment by Andarist** [comment]
Just to confirm - your git repo does have a master branch? Or is the base branch named differently?

**Comment by nategreen** [comment]
Yeah, there is still a master branch and nothing has changed.

Initially when we started seeing this, it happened to be on a branch created from another branch. We tried rebasing onto master, and also creating a fresh branch from master, each of which had the same results.

**Comment by Andarist** [comment]
Could you check if this is also failing for you locally? What's your git version on a failing machine? Could you try to run there `git merge-base master HEAD` and paste in the result?

**Comment by nategreen** [comment]
I did try running it locally in my initial efforts to debug/fix it. It worked fine locally, which was puzzling.

On the machine: `git version 2.29.2`.

The pipeline definition uses `checkout: self` to do the initial checkout of the repo. Here's an excerpt of the output from that checkout:

```
Syncing repository: <repo-name> (Git)
git version
git version 2.29.2
git lfs version
git-lfs/2.12.1 (GitHub; linux amd64; go 1.13.4)
git init "/home/vsts/work/1/s"
Initialized empty Git repository in /home/vsts/work/1/s/.git/
git remote add origin <repo-url>
git config gc.auto 0
git config --get-all http.<repo-url>.extraheader
git config --get-all http.proxy
git config http.version HTTP/1.1
git -c http.extraheader="AUTHORIZATION: bearer ***" fetch --force --tags --prune --progress --no-recurse-submodules origin
```

When I ran the command you said, it output:

```
fatal: Not a valid object name master
```

**Comment by nategreen** [comment]
Just out of curiosity, I tried `git merge-base main HEAD` (since `main` is now the default branch name in new repos)…that didn't work either. Same result: `fatal: Not a valid object name main`

**Comment by Andarist** [comment]
Are you sure that your setup was/is working correctly with the `@changesets/cli@2.12.0`? I would suspect that it just always return success from the status command (if it's not erroring).

I was able to reproduce your issue using the given git commands and they just don't lead to a proper checkout - is it, maybe, done afterward in your pipeline? Your cwd is pretty much empty, without any files in it (well, not counting `.git` directory). You are on a local master branch that has no commits on it. I would expect you to sync your empty master branch with the correct remote by doing smth like:
```
git pull -f origin master:master
```
This won't completely solve the issue yet - because the `status` compares the current HEAD to your base branch and you are not checked out on the correct commit yet! at this point. So I would expect that you'd run `changeset status` after doing a proper checkout.

**Comment by nategreen** [comment]
Maybe I missed some git activity between the steps I sent.

I do know that it was working before; it fails when there are no changesets in the PR branch and succeeds when they're added.

**Comment by Andarist** [comment]
Could u take a look and lost all git activities then? And also mention at what exact step changeset gets executed? It would help understanding the issue better.

**Comment by nategreen** [comment]
Here's the full log (with some details removed for brevity or security).

```
2021-01-21T14:31:01.8215392Z ##[section]Starting: Job
2021-01-21T14:31:02.0367887Z ##[section]Starting: Initialize job
2021-01-21T14:31:02.0369182Z Agent name: 'Hosted Agent'
2021-01-21T14:31:02.0369521Z Agent machine name: '<machine-name>'
2021-01-21T14:31:02.0369753Z Current agent version: '2.181.0'
2021-01-21T14:31:02.0412957Z ##[group]Operating System
2021-01-21T14:31:02.0413193Z Ubuntu
2021-01-21T14:31:02.0413351Z 18.04.5
2021-01-21T14:31:02.0413479Z LTS
2021-01-21T14:31:02.0413616Z ##[endgroup]
2021-01-21T14:31:02.0413784Z ##[group]Virtual Environment
2021-01-21T14:31:02.0413999Z Environment: ubuntu-18.04
2021-01-21T14:31:02.0414182Z Version: 20201210.0
2021-01-21T14:31:02.0414505Z Included Software: https://github.com/actions/virtual-environments/blob/ubuntu18/20201210.0/images/linux/Ubuntu1804-README.md
2021-01-21T14:31:02.0414814Z ##[endgroup]
2021-01-21T14:31:02.0415783Z Current image version: '20201210.0'
2021-01-21T14:31:02.0420601Z Agent running as: 'vsts'
2021-01-21T14:31:02.0467556Z Prepare build directory.
2021-01-21T14:31:02.0739327Z Set build variables.
2021-01-21T14:31:02.0773229Z Download all required tasks.
2021-01-21T14:31:02.0911622Z Downloading task: Bash (3.179.0)
2021-01-21T14:31:03.8212339Z Checking job knob settings.
2021-01-21T14:31:03.8220886Z    Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
2021-01-21T14:31:03.8221949Z    Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
2021-01-21T14:31:03.8222608Z Finished checking job knob settings.
2021-01-21T14:31:03.8486732Z Start tracking orphan processes.
2021-01-21T14:31:03.8676250Z ##[section]Finishing: Initialize job
2021-01-21T14:31:03.9041475Z ##[section]Starting: Checkout mri-design-system@debug-changesets-git to s
2021-01-21T14:31:03.9285217Z ==============================================================================
2021-01-21T14:31:03.9286098Z Task         : Get sources
2021-01-21T14:31:03.9286596Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2021-01-21T14:31:03.9286918Z Version      : 1.0.0
2021-01-21T14:31:03.9287307Z Author       : Microsoft
2021-01-21T14:31:03.9287905Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2021-01-21T14:31:03.9288271Z ==============================================================================
2021-01-21T14:31:04.8233671Z Syncing repository: <> (Git)
2021-01-21T14:31:04.8236629Z ##[command]git version
2021-01-21T14:31:04.8240553Z git version 2.29.2
2021-01-21T14:31:04.8241661Z ##[command]git lfs version
2021-01-21T14:31:04.8245550Z git-lfs/2.12.1 (GitHub; linux amd64; go 1.13.4)
2021-01-21T14:31:04.8247226Z ##[command]git init "/home/vsts/work/1/s"
2021-01-21T14:31:04.8250582Z Initialized empty Git repository in /home/vsts/work/1/s/.git/
2021-01-21T14:31:04.8252001Z ##[command]git remote add origin <repo-url>
2021-01-21T14:31:04.8259635Z ##[command]git config gc.auto 0
2021-01-21T14:31:04.8263802Z ##[command]git config --get-all http.<repo-url>.extraheader
2021-01-21T14:31:04.8267430Z ##[command]git config --get-all http.proxy
2021-01-21T14:31:04.8270796Z ##[command]git config http.version HTTP/1.1
2021-01-21T14:31:04.8277660Z ##[command]git -c http.extraheader="AUTHORIZATION: bearer ***" fetch --force --tags --prune --progress --no-recurse-submodules origin
2021-01-21T14:31:04.9133333Z remote: Azure Repos        
2021-01-21T14:31:05.6152460Z remote: 
2021-01-21T14:31:05.6158966Z remote: Found 17199 objects to send. (168 ms)        
2021-01-21T14:31:05.6166070Z Receiving objects:   0% (1/17199)
2021-01-21T14:31:05.6167140Z Receiving objects:   1% (172/17199)
...
2021-01-21T14:31:06.0758096Z Receiving objects: 100% (17199/17199), 15.08 MiB | 17.67 MiB/s, done.
2021-01-21T14:31:06.0787515Z Resolving deltas:   0% (0/9495)
2021-01-21T14:31:06.0887593Z Resolving deltas:   1% (95/9495)
2021-01-21T14:31:06.0899897Z Resolving deltas:   2% (190/9495)
...
2021-01-21T14:31:06.4974043Z Resolving deltas: 100% (9495/9495)
2021-01-21T14:31:06.4975032Z Resolving deltas: 100% (9495/9495), done.
2021-01-21T14:31:06.7738784Z From <repo-url>
2021-01-21T14:31:06.7831062Z  * [new branch]      Adding-kendo-grid-styles    -> origin/Adding-kendo-grid-styles
... (all branches)
2021-01-21T14:31:06.7937462Z  * [new branch]      validation-states           -> origin/validation-states
2021-01-21T14:31:06.7938996Z  * [new tag]         @mri/bootstrap-theme@0.2.0  -> @mri/bootstrap-theme@0.2.0

... (all tags)
2021-01-21T14:31:06.8210649Z  * [new tag]         @mri/theme-default@1.0.2-pre.0 -> @mri/theme-default@1.0.2-pre.0
2021-01-21T14:31:06.8211680Z  * [new tag]         v0.3.0-pre.0                -> v0.3.0-pre.0
2021-01-21T14:31:06.9030276Z ##[command]git -c http.extraheader="AUTHORIZATION: bearer ***" fetch --force --tags --prune --progress --no-recurse-submodules origin +f3ce...f72d:refs/remotes/origin/f3ce...f72d
2021-01-21T14:31:07.2093840Z From <repo-url>
2021-01-21T14:31:07.2095108Z  * [new ref]           f3ce...f72d -> origin/f3ce...f72d
2021-01-21T14:31:07.2541266Z ##[command]git checkout --progress --force f3ce...f72d
2021-01-21T14:31:07.2544990Z Note: switching to 'f3ce...f72d'.
2021-01-21T14:31:07.2545196Z 
2021-01-21T14:31:07.2545621Z You are in 'detached HEAD' state. You can look around, make experimental
2021-01-21T14:31:07.2546089Z changes and commit them, and you can discard any commits you make in this
2021-01-21T14:31:07.2546457Z state without impacting any branches by switching back to a branch.
2021-01-21T14:31:07.2546730Z 
2021-01-21T14:31:07.2546993Z If you want to create a new branch to retain commits you create, you may
2021-01-21T14:31:07.2547640Z do so (now or later) by using -c with the switch command. Example:
2021-01-21T14:31:07.2547977Z 
2021-01-21T14:31:07.2548852Z   git switch -c <new-branch-name>
2021-01-21T14:31:07.2549146Z 
2021-01-21T14:31:07.2553650Z Or undo this operation with:
2021-01-21T14:31:07.2553928Z 
2021-01-21T14:31:07.2554464Z   git switch -
2021-01-21T14:31:07.2554586Z 
2021-01-21T14:31:07.2554929Z Turn off this advice by setting config variable advice.detachedHead to false
2021-01-21T14:31:07.2555229Z 
2021-01-21T14:31:07.2555457Z HEAD is now at f3ce... Use latest @changesets/cli
2021-01-21T14:31:07.2561128Z ##[command]git config http.<repo-url>.extraheader "AUTHORIZATION: bearer ***"
2021-01-21T14:31:07.2651091Z ##[section]Finishing: Checkout mri-design-system@debug-changesets-git to s
2021-01-21T14:31:07.2676233Z ##[section]Starting: Bash
2021-01-21T14:31:07.2683847Z ==============================================================================
2021-01-21T14:31:07.2684164Z Task         : Bash
2021-01-21T14:31:07.2684430Z Description  : Run a Bash script on macOS, Linux, or Windows
2021-01-21T14:31:07.2684686Z Version      : 3.179.0
2021-01-21T14:31:07.2684936Z Author       : Microsoft Corporation
2021-01-21T14:31:07.2685272Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
2021-01-21T14:31:07.2685643Z ==============================================================================
2021-01-21T14:31:07.7705754Z Generating script.
2021-01-21T14:31:07.7706685Z Script contents:
2021-01-21T14:31:07.7707082Z git merge-base master HEAD
2021-01-21T14:31:07.7707880Z ========================== Starting Command Output ===========================
2021-01-21T14:31:07.7711337Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/4794...d708c.sh
2021-01-21T14:31:07.7711676Z fatal: Not a valid object name master
2021-01-21T14:31:07.7740150Z ##[error]Bash exited with code '128'.
2021-01-21T14:31:07.7764989Z ##[section]Finishing: Bash
2021-01-21T14:31:07.7786481Z ##[section]Starting: Checkout mri-design-system@debug-changesets-git to s
2021-01-21T14:31:07.7789906Z ==============================================================================
2021-01-21T14:31:07.7790186Z Task         : Get sources
2021-01-21T14:31:07.7790532Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2021-01-21T14:31:07.7790814Z Version      : 1.0.0
2021-01-21T14:31:07.7791000Z Author       : Microsoft
2021-01-21T14:31:07.7791405Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2021-01-21T14:31:07.7791770Z ==============================================================================
2021-01-21T14:31:08.0874175Z Cleaning any cached credential from repository: mri-design-system (Git)
2021-01-21T14:31:08.1078780Z ##[command]git version
2021-01-21T14:31:08.1414046Z git version 2.29.2
2021-01-21T14:31:08.1472981Z ##[command]git lfs version
2021-01-21T14:31:08.1585587Z git-lfs/2.12.1 (GitHub; linux amd64; go 1.13.4)
2021-01-21T14:31:08.1628292Z ##[command]git config --unset-all http.<repo-url>.extraheader
2021-01-21T14:31:08.1738150Z ##[section]Finishing: Checkout mri-design-system@debug-changesets-git to s
2021-01-21T14:31:08.1797942Z ##[section]Starting: Finalize Job
2021-01-21T14:31:08.1829274Z Cleaning up task key
2021-01-21T14:31:08.1830596Z Start cleaning up orphan processes.
2021-01-21T14:31:08.2159494Z ##[section]Finishing: Finalize Job
2021-01-21T14:31:08.2199125Z ##[section]Finishing: Job
```

**Comment by akphi** [comment]
> Are you sure that your setup was/is working correctly with the `@changesets/cli@2.12.0`? I would suspect that it just always return success from the status command (if it's not erroring).
> 
> I was able to reproduce your issue using the given git commands and they just don't lead to a proper checkout - is it, maybe, done afterward in your pipeline? Your cwd is pretty much empty, without any files in it (well, not counting `.git` directory). You are on a local master branch that has no commits on it. I would expect you to sync your empty master branch with the correct remote by doing smth like:
> 
> ```
> git pull -f origin master:master
> ```
> 
> This won't completely solve the issue yet - because the `status` compares the current HEAD to your base branch and you are not checked out on the correct commit yet! at this point. So I would expect that you'd run `changeset status` after doing a proper checkout.

@Andarist I have a github action that has a step calling `changeset status --since=master`. This works locally and fails in the pipeline. Your workaround really works. But I don't quite get your explanation. Apparently, when I log out `git status`, the output is `HEAD detached at pull/82/merge`. 

We just adopted `changesets` and we are looking for a way to fail the gate on each PR build if changeset is not present for all changed package.

**Comment by mitchellhamilton** [comment]
> I have a github action that has a step calling changeset status --since=master. This works locally and fails in the pipeline. Your workaround really works. But I don't quite get your explanation. Apparently, when I log out git status, the output is HEAD detached at pull/82/merge.

You most likely need to set `fetch-depth: 0` in `actions/checkout` so that it will include all the Git history

```yaml
- name: Checkout Repo
  uses: actions/checkout@main
  with:
    fetch-depth: 0
```

**Comment by akphi** [comment]
@mitchellhamilton yes, I did that and the error is still there. I wonder if you can see my pipeline. You can see the error there.
https://github.com/akphi/config-tester/pull/76/checks?check_run_id=2062957909

**Comment by Andarist** [comment]
@nategreen I don't quite see in this log the moment in which changesets are invoked 🤔 

@akphi it's quite hard to tell what has gone wrong - all refs should be downloaded in your case as per the action's source code:
https://github.com/actions/checkout/blob/25a956c84d5dd820d28caab9f86b8d183aeeff3d/src/git-source-provider.ts#L128-L134

Could you try running those on your action before you invoke changeset?
```
git merge-base master HEAD
git merge-base origin/master HEAD
git branch -r
```

**Comment by Andarist** [comment]
> Could this be due to the way checkout is done for PRs? Checkout step logs:

Definitely. I'm not a git expert though so usually I have to experiment quite a bit when it comes to such less known command/flags. 

As mentioned in the previous post - I'm still surprised that you both don't get your base branches fetched with `fetch-depth: 0` though as the source code of the checkout action explicitly mentions that it should be fetched. That being said - I think we can do better and fetch the missing git ref/timeline automatically for you but to assess what should be done I would really have to get more info about this. Could you try executing the mentioned (in the previous post) commands on your CI and paste in the results?

**Comment by akphi** [comment]
@rohit-gohri `fetch` won't do it, you need to `pull` - `git pull -f origin master:master`
@Andarist I have added the steps for the 4 commands you asked for:
```sh
git pull -f origin master:master # this needs to be run first else the other ones won't even show up
git merge-base master HEAD
git merge-base origin/master HEAD
git branch -r
```
https://github.com/akphi/config-tester/pull/76/checks?check_run_id=2067446422

**Comment by Andarist** [comment]
@akhpi so even `git branch -r` doesn't return anything if you don't make that `git pull -f origin master:master`?

@rohit-gohri you are right, I can find similar entries in @akphi's logs. This makes the whole thing even more strange

**Comment by Andarist** [comment]
@akphi could you push out a version without that extra `git pull -f origin master:master`? Can't find such on the list here: https://github.com/akphi/config-tester/actions

**Comment by akphi** [comment]
@Andarist So I did all 4 scenarios:
https://github.com/akphi/config-tester/pull/76

```
With `git pull`
No `git pull` and call `git merge-base master`
No `git pull` and call `git branch -r`
No `git pull` and call `git merge-base origin/master`
```

Looks like `git merge-base origin/master` is the good call here, should we then try to call `since=origin/master` instead?

**Comment by Andarist** [comment]
Ok, I think I somewhat know what to do for the @akhpi case - although the logs from git are somewhat inconsistent in a couple of places there.

I'm still not 100% sure what to do in the @nategreen's case - could you check what `git branch` and `git branch -r` returns for you there? I suspect that both lists are somewhat empty-ish but I'd like to confirm that first.

**Comment by tema1411** [comment]
Hello. I have thi problem too after run `yarn changeset`. I have master branch in my repository. 

**Error:**
```
�  error Error: Failed to find where HEAD diverged from master. Does master exist?
�  error     at getDivergedCommit (D:\myProject\styled-full-css-grid-2\node_modules\@changesets\git\dist\git.cjs.dev.js:58:11)
�  error     at processTicksAndRejections (internal/process/task_queues.js:93:5)
�  error     at async getChangedFilesSince (D:\myProject\styled-full-css-grid-2\node_modules\@changesets\git\dist\git.cjs.dev.js:169:22)
�  error     at async Object.getChangedPackagesSinceRef (D:\myProject\styled-full-css-grid-2\node_modules\@changesets\git\dist\git.cjs.dev.js:208:24)
�  error     at async add (D:\myProject\styled-full-css-grid-2\node_modules\@changesets\cli\dist\cli.cjs.dev.js:409:29)
�  error     at async run$1 (D:\myProject\styled-full-css-grid-2\node_modules\@changesets\cli\dist\cli.cjs.dev.js:1096:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

```json
//config.json
{
  "$schema": "https://unpkg.com/@changesets/config@1.5.0/schema.json",
  "changelog": "@changesets/cli/changelog",
  "commit": false,
  "linked": [],
  "access": "restricted",
  "baseBranch": "master",
  "updateInternalDependencies": "patch",
  "ignore": []
}
```

`git --version` // git version 2.22.0.windows.1
`yarn -v` //1.22.5
`node -v` //v14.15.5

**Comment by rohit-gohri** [comment]
Maybe we can log the actual error thrown by git here: https://github.com/atlassian/changesets/blob/eae72367e9cb8e708a1d441770dbc70930d4e1ad/packages/git/src/index.ts#L39-L44

`cmd.stderr` can be logged in debug mode?

**Comment by rohit-gohri** [comment]
This is what finally fixed it for me:

```yml
name: Node.js CI

on:
  push:
    branches: [develop, main]
  pull_request:

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v1
        with:
          node-version: 12

      - run: npm ci

      - run: |
          # Checkout PR head commit instead of merge commit
          # See: https://github.com/atlassian/changesets/issues/517
          git checkout ${{ github.event.pull_request.head.sha }}
          npx changeset status --since=origin/main
        if: ${{ github.event_name == 'pull_request' }}

```

Since according to https://github.community/t/running-git-merge-base-fork-point-master-command-as-part-of-a-job-fails/18212/2 `git merge-base` fails if the branch has already been merged before. And github pull_request events uses a merge commit to run the pipeline instead of the PR head. 

So I added a `git checkout ${{ github.event.pull_request.head.sha }}` command before running changeset.

**Comment by Andarist** [comment]
> git merge-base fails if the branch has already been merged before. And github pull_request events uses a merge commit to run the pipeline instead of the PR head.

wow, that's a nice finding! I think the whole thing can/should be just fixed on our side but I don't have much time to focus on this one right now properly to implement the fix.

**Comment by rohit-gohri** [comment]
I think `getChangedFilesSince` can be implemented like `danger-js` does it : https://github.com/danger/danger-js/blob/main/source/platforms/git/localGetDiff.ts

Use `git diff ${since}...HEAD` and then use [parse-diff](https://www.npmjs.com/package/parse-diff) to parse it and find modified files like in : https://github.com/danger/danger-js/blob/main/source/platforms/git/gitJSONToGitDSL.ts

**Comment by Andarist** [comment]
This could fix the issue mentioned in your comment but there are other ones mentioned in this thread and this would not fix them as in those cases "since" is an unknown ref (it has not been fetched at all).

**Comment by elpddev** [comment]
Thanks to the comment in this issue and others, I had success checking for changeset existence in a PR.

If I understood correctly, [github action checkout@v2](https://github.com/actions/checkout) only checkout a newly created merged commit that is the result of the PR branch head commit and the target commit. This throw off the `changeset status --from` [command](https://github.com/atlassian/changesets/blob/main/docs/automating-changesets.md#blocking) since the target commit/branch does not exists in the clone the github checkout action do.

To solve this:
1.  I fetch the origin PR commit and the base (PR target) commit. 
2. Afterward, I checkout to the PR commit because `changeset status` seems to not be able to compare the merged commit with the base commit. I did not investigate further. 
3. After running the changeset status check, I restore the current working directory and checkout the PR merged commit.

```
  git fetch origin ${GITHUB_HEAD_REF}
  git fetch origin ${GITHUB_BASE_REF}
  git checkout ${GITHUB_HEAD_REF}
  yarn changeset status --since=origin/${GITHUB_BASE_REF}
  git checkout ${GITHUB_SHA}
```

**Comment by elisechant** [comment]
In case this is helpful to others, this worked for me: 

```
git pull -f origin main:main
git merge-base main HEAD
git merge-base origin/main HEAD
git branch -r
yarn changeset status --since=main
```

**Comment by will-stone** [comment]
For anyone struggling with this in GitLab, I ended up using this in my job:

```
  script:
    # Required so that Changesets can compare with main branch
    - git config --global --add safe.directory /builds/path/to/project-name
    - git pull -f origin main:main
    - npx changeset status
```

Please let me know if there's any major downsides to this approach that I am not aware of 😬

**Comment by steve-taylor** [comment]
For those of you struggling with this in GitHub Actions, here's what works for me:

1. Check out the branch rather than the result of merging `main` into your branch. Also get the full commit history.
   ```yml
   uses: actions/checkout@v3
   with:
     ref: ${{github.event.pull_request.head.sha}}
     fetch-depth: 0
   ```
2. The workflow runner hasn't checked out `main` locally, so refer to it as `origin/main` in the `changeset status` command.
   ```yml
   run: npx changeset status --since origin/main
   ```

**Comment by glasser** [comment]
It seems like there are two different things that can cause the issue:
- git clone'ing in a way that doesn't fetch the `master` refer (to `origin/master`) at all
- git clone'ing all the refs, but not actually creating a local branch

If you're having the second problem (which I am, on CircleCI not GH Actions), it looks like there's a much simpler solution: just set `baseBranch` in `config.json` to `origin/master` instead of `master`.

As far as I can tell from a quick glance in the source, the only thing the tool ever actually does with `baseBranch` is pass it to `git merge-base BASE_BRANCH HEAD`, which can be any ref, not just a branch, so this seems fine.  (If it did things like check out the given baseBranch, that would be tougher.)

Perhaps `origin/main` would be a better default than `main`, even.

**Comment by kachkaev** [comment]
My solution for PR CI: https://github.com/blockprotocol/blockprotocol/pull/579

TL;DR:

```diff
       - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0 ## Needed for Changesets to find `main` branch

+      - run: |
+          git pull --force --no-tags origin main:main
+          yarn changeset status
```

Adding `--no-tags` to `git pull` makes command output less verbose, so Changesets report is surrounded with less noise. It’s also possible to forward the output of `git pull` to `/dev/null` but not sure if it’s worth it.

**Comment by clChenLiang** [comment]
Maybe just not have the local main branch

**Comment by alexbakers** [comment]
```bash
$ git merge-base master HEAD
fatal: Not a valid object name master
```

```bash
$ git pull -f origin master:master
fatal: couldn't find remote ref master
```

```bash
$ git merge-base origin/master HEAD
fatal: Not a valid object name origin/master
```

```bash
$ git branch -r
  origin/HEAD -> origin/develop
  origin/develop
```

**Comment by lsliwaradioluz** [comment]
I can confirm using

```bash
npx @changesets/cli status --since origin/main
```

worked for me on Github
